### DISCUSSION TITLE
Proposal to add configurable data to enhance Tasmota discovery

### DISCUSSION THREAD

**Discussion by blakadder** [arendst/Tasmota#19666] [Ideas]
## Proposal to add configurable data to Tasmota discovery with my focus being on Home Assistant

Considering there are now sensors being done in Berry and the increasing complexity of available sensors types the current solution is lacking some features.

Right now Tasmota discovery sends the payload of the tele//SENSOR topic to  tasmota/discovery/<id>/sensors which is then parsed in the Tasmota integration. This has a few shortcomings, especially if the sensor is not part of the main codebase and added with Berry:

1. there is no unit of measurement definition for each sensor which leaves the integration component guessing the unit from the available codebase or defaulting to a sensor that's not graphed historically in HA
2. all sensors are put in the main sensor category, you cant define a sensor as diagnostic or config config class
3. no ability to set binary sensor types unless they're switches, this is noticeable with the influx of presence sensors that don't have switch GPIO's
5. no ability to set sensor state_class so it is included in long term statistics

## Proposal

Have a supplemental discovery topic that allow defining additional attributes without complexity. The primary goal is easy implementation through get Berry and,  eventually, main code base. 

Format of choice would be JSON with some inspiration from rule triggers and Matter

The structure as follows:

```json
{"Presence":{"type":"binary", "class":"occupancy","icon":"mdi:house":},"BH1750#Illuminance":{"unit":"lux", "category":"diag", "class":"illuminance", "state_class":"measurement"},"SHTC3#Humidity":{"unit":"%", "category":"config", "class":"humidity", "state_class":"total"}}
```

Designed to follow the structure of rule trigger and matter sensor config allows you to choose a specific sensor and set key pair values with infinitely settable attributes independent of the platform. Key names for HA should be based on the existing MQTT discovery ones. 

This way makes eveerything optional and independent of the the existing discovery structure as well as easily expandable if new attributes are included in HA and could also include others like icons. 

This is primarily targeted at ESP32 where there's more potential to have custom sensors and other types of peripherals

**Comment by sfromis** [comment]
@emontnemery  Did you see this suggestion?

Questions about being able to supply units for measurements from Tasmota sensor drivers implemented in Berry are popping up from time to time, about drivers not becoming part of the Tasmota project. Instead of directing people to make suggestions for https://github.com/emontnemery/hatasmota/ it would make sense to have the driver being able to provide such metadata.

I was thinking about a simpler approach, like a driver giving `"MySensor":{"MyValue":123,"AnotherValue":456}` and then have the same payload also include metadata in the style of the existing `TempUnit`, `PressureUnit` and `SpeedUnit`, but with a naming convention not needing the relationships hardcoded: `"MyValue_unit":"Parsec","AnotherValue_unit":"Bananas"` or `"Units":{"MyValue":"Parsec","AnotherValue":"Bananas"}`. This assumes that multiple drivers would be able to "agree" on same units for same names.

Of course, I assume that the more general format suggested by blakadder would be even more useful as a reasonably complete solution, albeit with more infrastructure needed both ends to carry an additional discovery payload topic.

**Comment by arendst** [comment]
Here's another approach.

Current report is this:
```
"HTU21":{"Temperature":23.5,"Humidity":43.2,"DewPoint":10.2,"HeatIndex":23.0}
```
Adding units needs a redesign adding a sublevel like:
```
"HTU21":{"Temperature":{"value":23.5,"unit_of_measurement":"°C"},Humidity":{"value":43.2,"unit_of_measurement":"%"},"DewPoint":{"value":10.2,"unit_of_measurement":"°C"},"HeatIndex":{"value":23.0,"unit_of_measurement":"°C"}}
```
This becomes a book and adding more sensors in the same JSON is a nightmare.
A shorter notation would be:
```
"HTU21":{"Temperature":{"v":23.5,"u":"°C"},Humidity":{"v":43.2,"u":"%"},"DewPoint":{"v":10.2,"u":"°C"},"HeatIndex":{"v":23.0,"u":"°C"}}
```

**Reply by sfromis** [reply]
While those object hierarchies make sense from a data modeling point of view, they'd be a breaking change to users of the payloads. It would be "reasonably backwards compatible" with a structure like:
```
"HTU21":{"Temperature":23.5,"Temperature_unit":"°C","Humidity":43.2,"Humidity_unit":"%","DewPoint":10.2,"DewPoint_unit":"°C","HeatIndex":23.0,"HeatIndex_unit":"°C"}}
```
... or the already mentioned structures keeping the metadata as added data instead of breaking existing schemas. While I'm not terribly fond of naming-based linking, it could be an add-on.

Or the structure I mentioned with all the units in a separate `Units` object, with the drawback of this being more complicated to handle both ends. The benefit would be to avoid linking by names, and that multiple sensors having the same keys could share units (I do not see much use case for some temperatures in C and others in F).

(Arguably, the unit for humidity is `RH` or `%RH`, but that's another discussion).

**Reply by barbudor** [reply]
Instead of modifying the SENSOR message and make it heavier at each publucation why not improving the discovery message
Instead of including dummy value, the discovery message could give the unit
Of course not changing (for now) the existing 'Unit' for temperature as existing today to not break existing use cases.
It may be requiring a new call back in each sensor driver to provide those units
